### ISSUE TITLE
[REVIEW]: diwanalr: An R data analysis package for diffusing-wave spectroscopy

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#947]
**Submitting author:** @peterjwatkins (<a href="http://orcid.org/0000-0002-1290-3843">Peter Watkins</a>)
**Repository:** <a href="https://github.com/peterjwatkins/diwanalr" target ="_blank">https://github.com/peterjwatkins/diwanalr</a>
**Version:** v0.1
**Editor:** @arfon
**Reviewer:** @alexpghayes
**Archive:** <a href="https://doi.org/10.5281/zenodo.3247971" target="_blank">10.5281/zenodo.3247971</a>

## Status

[![status](http://joss.theoj.org/papers/a5f7da635b799f4befef712ec0a59127/status.svg)](http://joss.theoj.org/papers/a5f7da635b799f4befef712ec0a59127)

Status badge code:

```
HTML: <a href="http://joss.theoj.org/papers/a5f7da635b799f4befef712ec0a59127"><img src="http://joss.theoj.org/papers/a5f7da635b799f4befef712ec0a59127/status.svg"></a>
Markdown: [![status](http://joss.theoj.org/papers/a5f7da635b799f4befef712ec0a59127/status.svg)](http://joss.theoj.org/papers/a5f7da635b799f4befef712ec0a59127)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/peterjwatkins/diwanalr" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@alexpghayes, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.theoj.org/about#reviewer_guidelines. Any questions/concerns please let @arfon know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/peterjwatkins/diwanalr">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** Does the release version given match the GitHub release (v0.1)?
- [x] **Authorship:** Has the submitting author (@peterjwatkins) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @alexpghayes it looks like you're currently assigned as the reviewer for this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.00947/joss.00947/10.21105.joss.00947.pdf)

**Comment by arfon** [comment]
👋 @alexpghayes - please take a look at the review checklist above and work your way through it. If you have any questions for @peterjwatkins as you go, ask them here or make issues on the repository: https://github.com/peterjwatkins/diwanalr

If anything isn't clear, just give me a shout!

**Comment by alexpghayes** [comment]
## Summary

This package appears to perform a set of useful calculations, but is not yet ready for public consumption.

### Documentation

- I recommend adding the material in the intro notebook / vignette to the Github README. An increasing number of people see the README before anything else. Most R users don't use Jupyter notebooks, so it doesn't make sense for the most public facing introduction to be in a Jupyter notebook.

- I recommend printing the original data and the intermediate data frames throughout the intro (this will happen automatically if you use a `.Rmd` README, which you can setup with `usethis::use_readme_rmd()`). This will help users understand the input and output of the `diwanalr` functions. Similarly, having plots in the README will make it more compelling.

- Single letter variables names are uninformative and considered bad style in R. I would give variables throughout the documentation meaningful names.

 - `calc_g1()` uses a smoothing spline, but based on my reading of the documentation I had expected it to make an exact calculation.

### Function documentation

Some questions I had reading the documentation for various functions:
- `calc_g1()`: What is `n`? What does it do? Do the columns in `d` need to be named? Do they need to be in a particular order?
- `plot_g1()`: What constitutes an acceptable input tibble?
- `calc_modulus`: What are the units of `radius`?
 
### Provided examples

I get some warnings running the examples in the vignette:

``` r
library(diwanalr)

data(dws)
d <- calc_g1(dws)
gp <- plot_g1(d) ; gp
e <- calc_msd(d)
ep <- plot_msd(e) ; ep
f <- calc_modulus(e)
fp <- plot_modulus(f) ; fp  # gives a warning for me
#> Warning in self$trans$transform(x): NaNs produced
#> Warning: Transformation introduced infinite values in continuous y-axis
#> Warning: Removed 165 rows containing missing values (geom_point).
```

Created on 2018-09-18 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

### R package development practice

- Variables in examples and code should have informative names, rather than just being a single letter
- Function arguments should also have informative names
- Depending on the tidyverse is bad practice and susecptable to breakage -- https://www.tidyverse.org/articles/2018/06/tidyverse-not-for-packages/

### Tests

- The existing tests check the type of objects returned from functions, but they should also check the correctness of the calculations for some simple cases

- The package should pass R CMD Check with no ERRORS or WARNINGS and ideally no NOTES. To run R CMD Check, install the `devtools` package and then run `devtools::check()`. More details are [available here](http://r-pkgs.had.co.nz/).

- The tests should be automated via some sort of continuous integration service. `usethis::use_travis()` can help set this up.

### Community guidelines

- The are no community/contribution guidelines.
- `usethis::use_tidy_contributing()` can help set this up.

### References

I'm not sure what needs a DOI or how to check that one is present.

**Comment by alexpghayes** [comment]
cc @peterjwatkins @arfon -- do I need to tag y'all to make sure you get notifications or is just posting to this thread enough?

**Comment by arfon** [comment]
> cc @peterjwatkins @arfon -- do I need to tag y'all to make sure you get notifications or is just posting to this thread enough?

We all got tagged at the top by @whedon so @peterjwatkins should be getting these notifications :-)

**Comment by peterjwatkins** [comment]
@peterjwatkins is getting the notifications :-)

Thanks for the review and feedback.. about to go on leave for 3 weeks, and I'll get back when I return.

**Comment by peterjwatkins** [comment]
Just to advise that I've been away for a little while, and I'm working with the comments from @alexpghayes

**Comment by peterjwatkins** [comment]
Thanks @alexpghayes for the review/feedback in relation to the package.

Some preliminary comments in reply to the review/feedback.

In relation to the documentation, I have made the changes in relation to the documentation; I've added most of the vignette material to the README which, as a result, now has the plots embedded. It does look a lot more compelling. The comment in relation to single letter variables is noted; these have been removed and replaced by more meaningful names, which will be more understandable to the reader. This applies to the function names, along with the arguments to same.

I have clarified the use of the spline approach in the calculation as well.

In relation to the examples, the warnings appear to the presence of negative numbers within the data, and the use of log transformation on the graph scale generates these messages. I have used 'suppressWarnings' in the function to remove these.

In the development practice, I have removed the generic call to the 'tidyverse' and have made explicit calls to the related package needed for the respective functions.

With respect to the tests, the correctness of the calculations needs the use of a standarised set of solutions to use with the DWS technique. This is a work presently in progress.

As suggested, I've been using 'devtools::check()' and there are two issues which are arising; one relates to
checking examples ... 
*ERROR
  Running examples in 'diwanalr-Ex.R' failed
  The error most likely occurred in:
  
  > base::assign(".ptime", proc.time(), pos = "CheckExEnv")
  > ### Name: calc_freq
  > ### Title: This function is used to calculate the frequency as inverse of
  > ###   time
  > ### Aliases: calc_freq
  > 
  > ### ** Examples
  > 
  > calc_freq(time)
  Error in calc_freq(time) : could not find function "calc_freq"*

which *is* present in the code,

and
*library("testthat")
Error in library("testthat") : there is no package called 'testthat'

  Execution halted*

even though 'testthat' is present.

I'll address the remaining comments subsequently.

**Comment by peterjwatkins** [comment]
@arfon I have responded to the comments made by @alexpghayes

**Comment by arfon** [comment]
@alexpghayes - could you come and take another look when you get a chance?

**Comment by alexpghayes** [comment]
I've made some additional comments in peterjwatkins/diwanalr#1, where I did a bit of code cleanup for R CMD Check.

**Comment by arfon** [comment]
@peterjwatkins - I think this might be ready for you to take another look.

**Comment by arfon** [comment]
@peterjwatkins @alexpghayes - how are we doing here?

**Comment by peterjwatkins** [comment]
?Hi guys,


I've been lagging on this one - work was busy at the end of last year, and presently on leave. I'll tidy this up in the next few weeks.


Rgds Peter

________________________________

**Comment by peterjwatkins** [comment]
@alexpghayes<https://github.com/alexpghayes>

I have had the chance to revisit this, and thee’s an issue with travis with running the examples.

[cid:image002.jpg@01D4B7DF.AAD0AEE0]

I’m not certain where to find ‘diwanalr-Ex.R’ file to sort this error (first time at doing this).. are there any suggestions at all?

Thanks

**Comment by alexpghayes** [comment]
This means the code in one of the examples isn't self-contained and/or breaks somehow. I recommend using `devtools::run_examples()` to figure out which example is broken.

**Comment by peterjwatkins** [comment]
@alexpghayes - I've cleaned up the code and R CMD runs OK now, with only 1 note now.. can you review when convenient.

**Comment by alexpghayes** [comment]
Can you comment on:

https://github.com/peterjwatkins/diwanalr/blob/alexpghayes-r_cmd_cleanup_master/R/msd.R#L51

and 

https://github.com/peterjwatkins/diwanalr/blob/alexpghayes-r_cmd_cleanup_master/R/modulus.R#L110

Why are `NA` values being generated and why is it safe to ignore them?

At this point the code is decently clean but hitting it with [`styler`](https://github.com/r-lib/styler) might make it slightly more readable.

The tests currently only check that calculations are in a `data.frame()`. They should really verify that the calculated values matched with some expected values for easy tests cases. So far it looks like the package was developed on a single dataset, so trying the functions on other datasets would be a good idea (as would using these additional datasets for additional tests).

**Comment by peterjwatkins** [comment]
Thanks @alexpghayes for the feedback.

I've added comments to the files (to make it explicit for the code readers) about these lines -
re: msd.R#L51, I've added as explanations:
Note : NAs are introduced in the 'findX' function and the last statement is used as a filter for the NAs's
while for modulus.R#L110, I've added:
Negative numbers are generated, which are artefacts from the calculation which generates warnings when this function is deployed, and is not serious. The last statement is used to filter out these statements.

The main R code has also been prettified using the 'styler' package as well.

The last point is noted, and is accurate. There is only one dataset  present in the package, used to develop the package. Other datasets are being/will be generated, that will be included with further/future package development. This is a work in progress. Thanks for the feedback.

**Comment by alexpghayes** [comment]
Quick note on comments:

> Note : NAs are introduced in the 'findX' function and the last statement is used as a filter for the NAs's

This is useful in that it describes *what* is happening, but it doesn't describe *why* it is happening. If someone read this code and was trying to decide if this where a reasonable decision, they'd have to go back and do a bunch of math themselves. I recommend describing exactly which inputs get turned into `NA`s and why it is safe to ignore that input. Without understanding the methods being used here, this feels a lot like throwing out inconvenient data, which makes me suspicious. Some more detailed comments about this could help users trust the software more. Same goes for the second comment.

**Comment by peterjwatkins** [comment]
OK, got it now :-)

In relation to the first comment, the introduction of NA relates to 'uniroot'; NA is introduced when there is no solution for 'uniroot' - I've added the following
  ## For 'uniroot' to behave properly, it's important that f(lower) * f(upper) < 0
  ## This test ensures that a solution is present to find X
  ## Otherwise, there is no root and thus NA
which explains why NA is introduced, and then subsequently filtered.

In relation to the second comment; I've revised the comments to:
  ## NaNs are generated when plotting modulus due to negative numbers
  ## in G' and G'', which causes warnings with ggplot.
  ## These are artefacts, and have no impact on the visualisation in the plot
This function is intended as a tool for the end package user to visualise the calculated G' and G''. Inherently, there are negative numbers computed for these values which causes the warnings to appear, without suppression. From a coding viewpoint, I acknowledge the point but the users end, this tool is used for data visualisation, and later use.

**Comment by arfon** [comment]
@alexpghayes - does this sound reasonable ☝️?

**Comment by alexpghayes** [comment]
Will circle back to this review later this week

**Comment by alexpghayes** [comment]
Much improved! For the comment

```
 For 'uniroot' to behave properly, it's important that f(lower) * f(upper) < 0.
 This test ensures that a solution is present to find X. Otherwise, there is no root and thus NA
```

it's worth adding whether or not we expect a solution to exist, and why. In both cases, it's worth interpreting what this means in terms of the scientific data in addition to the math.

At https://github.com/peterjwatkins/diwanalr/blob/alexpghayes-r_cmd_cleanup_master/R/modulus.R#L110 we have

```
  ## NaNs are generated when plotting modulus due to negative numbers
  ## in G' and G'', which causes warnings with ggplot.
  ## These are artefacts, and have no impact on the visualisation in the plot
```

It looks like modulus values are being plotted directly, and then when the modulus is negative the `scale_*_log()` is breaking because we try to take the log of a negative value. If ignoring negative modulus is fine, it's worth adding a note on this, because otherwise this is just throwing away data. Do we expect modulus to always be nonnegative and then numerical error sometimes makes it negative?

**Comment by alexpghayes** [comment]
Beyond this, I'm happy with the current state of the package with the outstanding issue of correctness tests.

**Comment by arfon** [comment]
Excellent, thanks @alexpghayes. @peterjwatkins - please let us know here when you've had a chance to incorporate @alexpghayes' feedback.

**Comment by labarba** [comment]
@peterjwatkins 👋  — it looks like we're waiting on some improvements on your side? Let us know your status, when you have a chance.

**Comment by peterjwatkins** [comment]
@labarba - apologies for the delay in responding - I will attend to this over the next few days

**Comment by labarba** [comment]
@peterjwatkins — Can you give us a status update?

**Comment by peterjwatkins** [comment]
@labarba - here's the update :)

For the plot_modulus function, I have revised the preliminary comments to
  ## Filter (> y_threshold) is used for modulus for data visualisation
  ## Modulus values <=0 create NaNs, displaying error/warning .
  ## This results in only positive modulus values being displayed, and not
  ## negative values. For the purpose of visualisation, this is satisfactory since 
  ## only positive values are of interest; no data loss occurs with this function.
which, I believe, satisfies the second remark from Alex.

In relation to the first comment, I'm not able to respond to this one as requested. The DWS technique is not my area of expertise; my involvement in this area was developing the software so I can't respond whether a solution exists or otherwise in these cases. My guess is that this would also dependent on the actual systems are being studied using the DWS technique, and will become more evident with further datasets. Consequently, I'm not able to provide any further revision in this case.

Alex also notes that there's an issue relating to tests of completeness, needing simpler datasets. This is still a work in progress with this data to be collected for testing. Thus, I acknowledge that there is a deficit in this case but this will be revised and updated with future revisions of the package.

**Comment by alexpghayes** [comment]
@peterjwatkins If you don't have real world data sets to use in tests, you can simulate fake data and use the fake data to confirm that calculations are correct.

@labarba Do I make an accept / reject recommendation at some point or just evaluate if the review checklist items have been completed?

**Comment by labarba** [comment]
@alexpghayes Once you are satisfied with the author's responses or code modifications, we do ask that you state here that you recommend publication. Thanks!

**Comment by peterjwatkins** [comment]
At present, I am developing test data for testing the calculations.

**Comment by peterjwatkins** [comment]
@alexpghayes - thanks for the suggestion.
To that end, a model response was used for the tests. An acknowledgement of the source has been added to 'testthat.R' as a comment (see below). This replaces the use of 'data(dws)' used in previous testing.

"Test performed on the calculations use model data.
The model response is taken from: Coarsening and Rheology of Casein and Surfactant Foams
A. Saint-Jalmes, S. Marze and D. Langevin, Equations 4 and 5, page 276
Food Colloids: Interactions, Microstructure and Processing, ed. by Eric Dickinson
g2(t) ~ 1 + beta * g1(t)^2 and g1(t) ~ exp(-2sqrt(6t/tm))

-------- R code to produce model data
span <- seq(-8, 1, 0.05) ; time <- 10^span
t_half <- 0.05
g2 <- exp(-2 * sqrt(6 * time/t_half))
g1 <- 1 + 0.9 * g2^2
test <- tibble::as_tibble(cbind(time, g2))
----------

Tests have been added for the _g1_(_t_), msd and modulus calculations in relation to the calculations.

**Comment by labarba** [comment]
@alexpghayes — Looks like we're waiting for you to comment on the author's reply from 10 days ago. Let us know your assessment.

**Comment by alexpghayes** [comment]
Notes on the tests:

- https://github.com/peterjwatkins/diwanalr/blob/alexpghayes-r_cmd_cleanup_master/tests/testthat/test_g1.R#L10 does not check any code in the package itself. In general tests should call package functions.
- The tolerance for the tests is very high (0.1 or 0.01 instead of the typically machine precision of around 1e-16. Even loose tests in the numeric world still use tolerances around 1e-6 to 1e-8). Can you comment on why this is the case?
- https://github.com/peterjwatkins/diwanalr/blob/alexpghayes-r_cmd_cleanup_master/tests/testthat/test_g1.R#L10 checks only three rows out of a much larger data frame. It would be better to test against all the values in the data frame. Similarly at https://github.com/peterjwatkins/diwanalr/blob/alexpghayes-r_cmd_cleanup_master/tests/testthat/test_msd.R#L27.

Once these are addressed I'm happy to recommend publication.

**Comment by peterjwatkins** [comment]
In relation to the notes:

- The tests for test_g1.R have been revised. The tests are performed on data generated on the model data, rather than specific calculations as was done previously.

- As noted, the tolerance is high and it is also been revised to 0.05. There is considerable variation in the  result (1.00 +/- 0.05) which is calculated using normalisation of the original data. Using typical machine precision would cause the test to fail, even with precision values of 1e-6 to 1e-8. My aim was to demonstrate the calculations provide a result to close to 1.00 for the early part of the data and close to 0.00 for the latter part of the data. I am satisfied that these tests are suitable and fit-for-purpose. In the event where higher precision is needed for testing the calculations, then I acknowledge and agree that more stricter tolerances would be needed (as demonstrated in test_msd.R, for example).
- The comment is noted but I believe that these tests are sufficient. The tests are meant to identify whether the calculation is providing the correct result. The tests are passing which indicates that the result lies within the required precision for these values, and thus are fit-for-purpose and also sufficient. In this case, I feel that by adding more would not add further value to the purpose of testing of the package.

**Comment by alexpghayes** [comment]
Yep! Happy to recommend for publication!

**Comment by peterjwatkins** [comment]
Thanks @alexpghayes !

**Comment by arfon** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.00947/joss.00947/10.21105.joss.00947.pdf)

**Comment by arfon** [comment]
Hi @peterjwatkins - we're nearly ready to accept here. Before we do so, could you make a few updates to your paper:

- If possible, I think it might read better if the large equation was broken out by itself.
- Would it be possible to expand a little on the introduction/statement of need section? We ask that authors include in the paper some sentences that would explain the software functionality and domain of use to a non-specialist reader.
- Are there examples of existing academic work that make use of this package? If so, could you cite them here?

**Comment by peterjwatkins** [comment]
Hi @arfon, as requested, I've made updates to the paper:

- the large equation has been separated and also centered
- I've added some further detail to the introduction as follows:
_____________________________
DWS is an optical technique derived from dynamic light scattering which studies the dynamics of scattered light [Weitz].

If carefully calibrated, DWS allows the quantitative measurement of microscopic motion in a soft material, which can be used with micro-rheology to determine the rheological properties of such a complex medium.

DWS has been applied the analysis to food systems. Some examples include understanding the dynamics of emulsion  systems (e.g. fat in water) such as mayonnaise [Kim 2019] along with the gellation of food products [Alexander 2006].

DWS also has practical applications in the pharmaceutical and cosmetic industries as well [@Reufer2014].

In some instances, the analysis of data resulting from DWS measurements can be tedious depending on, for example, the use of spreadsheets for numerical analysis, or multiple software packages [Niederquell 2017]. The **diwanalr** package has been developed as an alternative approach, allowing data analysis to be readily performed using a single software package, within R. The package has a set of functions which can be readily and seamlessly deployed after the measurement of data. Further functionality can be added to the package as required.
____________________________________

- Unfortunately, there is no publically available work at this stage which uses the package, so I am unable to include any citations at this stage.

I trust that this is satisfactory.

**Comment by arfon** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.00947/joss.00947/10.21105.joss.00947.pdf)

**Comment by arfon** [comment]
Thanks @peterjwatkins. That alignment didn't quite work properly - but this fix makes things good I think: https://github.com/peterjwatkins/diwanalr/pull/2

**Comment by peterjwatkins** [comment]
Thanks @arfon -  as an FYI, I pulled the fix down into R, and under preview, the equation does not display. However, it would be better to compile it as a PDF and see if it works  it might be just a quirk.

**Comment by arfon** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.00947/joss.00947/10.21105.joss.00947.pdf)

**Comment by arfon** [comment]
@peterjwatkins ☝️this looks good to me. What about you? Although I failed to add a space before the reference (fixed in https://github.com/peterjwatkins/diwanalr/pull/3)

**Comment by peterjwatkins** [comment]
@arfon the only minor thing is that the first 'g' is not rendered correctly - instead of _g_ it is *g* - might be better to remove the '*' - only a small detail

**Comment by arfon** [comment]
> @arfon the only minor thing is that the first 'g' is not rendered correctly - instead of g it is * g * - might be better to remove the '*' - only a small detail

https://github.com/peterjwatkins/diwanalr/pull/3 fixes this too now I think.

**Comment by peterjwatkins** [comment]
@arfon OK.. good to go then :-)

**Comment by arfon** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.00947/joss.00947/10.21105.joss.00947.pdf)

**Comment by arfon** [comment]
@peterjwatkins could you merge https://github.com/peterjwatkins/diwanalr/pull/3 please?

**Comment by arfon** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.00947/joss.00947/10.21105.joss.00947.pdf)

**Comment by arfon** [comment]
@peterjwatkins - At this point could you make a new release of this software that includes the changes that have resulted from this review. Then, please make an archive of the software in Zenodo/figshare/other service and update this thread with the DOI of the archive? For the Zenodo/figshare archive, please make sure that:

- The title of the archive is the same as the JOSS paper title
- That the authors of the archive are the same as the JOSS paper authors

I can then move forward with accepting the submission.

**Comment by peterjwatkins** [comment]
@arfon - I've released the software and it has been archived on Zenodo,  and the DOI is 10.5281/zenodo.3427971

**Comment by arfon** [comment]
Hrm, looks like that DOI isn't resolving yet (http://doi.org/10.5281/zenodo.3427971), can you confirm you've hit the 'publish' button on Zenodo?

**Comment by peterjwatkins** [comment]
@arfon.. it turns out, I hadn't :-) - first time that I have used Zenodo.. I've edited the page and hit the 'publish' button on Zenodo.. not certain on how long it will take for the DOI to resolve.

**Comment by arfon** [comment]
> @arfon.. it turns out, I hadn't :-) - first time that I have used Zenodo.. I've edited the page and hit the 'publish' button on Zenodo.. not certain on how long it will take for the DOI to resolve.

Thanks! It's still not resolving but I'm sure it will soon. I'll set myself a reminder.

**Comment by arfon** [comment]
@whedon remind @arfon in 1 day

**Comment by whedon** [comment]
Reminder set for @arfon in 1 day

**Comment by arfon** [comment]
@peterjwatkins it looks like http://doi.org/10.5281/zenodo.3427971 still isn't resolving. I think you might need to contact Zenodo support to find out what's going on here.

**Comment by peterjwatkins** [comment]
@arfon - I've realised what the issue is - I made a transcription error - the correct URL is http://doi.org/10.5281/zeonodo.3247971 - this is correctly resolving

**Comment by peterjwatkins** [comment]
https://doi.org/10.5281/zenodo.3247971

**Comment by arfon** [comment]
@whedon set 10.5281/zenodo.3247971 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.3247971" target="_blank">10.5281/zenodo.3247971</a> is the archive.

**Comment by arfon** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- 10.1073/pnas.1817029116 is OK
- https://doi.org/10.1016/j.foodhyd.2005.02.021 is OK
- https://doi.org/10.1002/jps.24197 is OK
- https://doi.org/10.1016/j.ijpharm.2017.07.038 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/790

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/790, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by arfon** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.00947/joss.00947/10.21105.joss.00947.pdf)

**Comment by arfon** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1142955236702769154) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/791
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.00947](https://doi.org/10.21105/joss.00947)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? notify your editorial technical team...

**Comment by arfon** [comment]
@alexpghayes - many thanks for your review here ✨ 

@peterjwatkins - your paper is now accepted into JOSS :zap::rocket::boom:

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](http://joss.theoj.org/papers/10.21105/joss.00947/status.svg)](https://doi.org/10.21105/joss.00947)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.00947">
  <img src="http://joss.theoj.org/papers/10.21105/joss.00947/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: http://joss.theoj.org/papers/10.21105/joss.00947/status.svg
   :target: https://doi.org/10.21105/joss.00947
```
This is how it will look in your documentation:

[![DOI](http://joss.theoj.org/papers/10.21105/joss.00947/status.svg)](https://doi.org/10.21105/joss.00947)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: http://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.salsalabs.org/donate-to-joss

**Comment by peterjwatkins** [comment]
Thanks @arfon and @alexpghayes, it's been a great learning experience.
